docs/docs/adrs/adr-015-partial-set-security.md

```diff
@@ -0,0 +1,300 @@
+
+
+If a proposal has those arguments wrongly set, it should get rejected in [ValidateBasic](https://github.com/cosmos/interchain-security/blob/v4.0.0/x/ccv/provider/types/proposal.go#L86).
+
+In the code, we distinguish whether a chain is _Top N_ or _Opt In_ by checking whether `top_N_fraction` is set or not.
+
+In a future version of PSS, we intend to introduce a way to modify the parameters of a consumer chain, e.g, a chain that is _Opt In_ to become _Top N_, etc.
+
+#### State & Query
+We augment the provider module’s state to keep track of the `top_N_fraction` value for each consumer chain. The key to store this information would be:
+
+```
+topNFractionBytePrefix | len(chainID) | chainID
+```
+To create the above key, we can use [`ChainIdWithLenKey`](https://github.com/cosmos/interchain-security/blob/v4.0.0/x/ccv/provider/types/keys.go#L418).
+
+Then in the [keeper](https://github.com/cosmos/interchain-security/blob/v4.0.0/x/ccv/provider/keeper/keeper.go) we introduce methods as follows:
+```golang
+func (k Keeper) SetTopN(ctx sdk.Context, chainID string, topNFraction string)
+func (k Keeper) IsTopN(ctx sdk.Context, chainID string) bool
+func (k Keeper) IsOptIn(ctx sdk.Context, chainID string) bool
+
+// returns the N if Top N chain, otherwise an error
+func (k Keeper) GetTopN(ctx sdk.Context, chainID string) (uint8, error)
+```
+
+We also extend the `interchain-security-pd query provider list-consumer-chains` query to return information on whether a consumer chain is an Opt In or a Top N chain and with what N. This way, block explorers like [Mintscan](https://www.mintscan.io/) could show for a chain “This chain is secured by N% of the provider chain.”
+
+### How do validators opt in?
+A validator can opt in by sending a new type of message that we introduce in [tx.proto](https://github.com/cosmos/interchain-security/blob/v4.0.0/proto/interchain_security/ccv/provider/v1/tx.proto#L1).
+```protobuf
+message MsgOptIn {
+    // the chain id of the consumer chain to opt in to
+    string chain_id = 1;
+    // the provider address of the validator
+    string provider_addr = 2;
+    // (optional) the consensus public key to use on the consumer
+    optional string consumer_key = 3;
+}
+```
+Note that in a Top N consumer chain, the top `N%` provider validators have to validate the consumer chain.
+Nevertheless, validators in the bottom `(100 - N)%` can opt in to validate as well.
+Provider validators that belong to the top `N%` validators are immediately opted in to validate a Top N consumer chain.
+This means that if a validator `V` belongs to the top `N%` validators but later falls (e.g., due to undelegations) to the bottom `(100 - N)%`, `V` would still be considered opted in and has to validate unless `V` sends a `MsgOptOut` message (see below).
```

Comment:
So you would have to opt out again every time you get in the top N, right?
i.e.
val V is in top N
falls below top N
opts out
gets back in top N
falls below again
has to opt out again

not really a problem, because this is already annoying and validators that skirt in and out of top N probably also don't really want to turn on/off their nodes all the time, so they'd probably just stay opted in until they are for sure not in the top N anymore.

Reply:
cc-ing @jtremback in regards to the later communication part